src/util/timer.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::{max, Ordering};
+use std::time::Duration;
+use util::time::Instant;
+use std::collections::BinaryHeap;
+
+struct TimeoutTask<T> {
+    next_tick: Instant,
+    timeout: Duration,
+    task: T,
+}
+
+impl<T> PartialEq for TimeoutTask<T> {
+    fn eq(&self, other: &TimeoutTask<T>) -> bool {
+        if self.next_tick == other.next_tick {
+            return true;
+        }
+        false
+    }
+}
+
+impl<T> Eq for TimeoutTask<T> {}
+
+impl<T> PartialOrd for TimeoutTask<T> {
+    fn partial_cmp(&self, other: &TimeoutTask<T>) -> Option<Ordering> {
+        Some(if self.next_tick > other.next_tick {
+            Ordering::Less
+        } else if self.next_tick < other.next_tick {
+            Ordering::Greater
+        } else {
+            Ordering::Equal
+        })
+    }
+}
+
+impl<T> Ord for TimeoutTask<T> {
+    fn cmp(&self, other: &TimeoutTask<T>) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+pub struct Timer<T> {
+    now: Instant,
+    tick: Duration,
+    pending: BinaryHeap<TimeoutTask<T>>,
+    delivered: Vec<TimeoutTask<T>>,
+}
+
+impl<T> Timer<T> {
+    fn align_instant(&self, ins: Instant) -> Instant {
+        let mut next = self.now + self.tick;
+        loop {
+            if next < ins {
+                next += self.tick;
+                continue;
+            }
+            break next;
+        }
+    }
+
+    /// Create a new timer with a specified `tick` which means
+    /// the min timeout when the `Timer` calls `next_timeout`.
+    pub fn new(tick: Duration) -> Self {
+        Timer {
+            now: Instant::now_coarse(),
```

Comment:
The resolution of `now_coarse` is millisecond. The granularity may scales from 1ms to 10ms based on the different config of the kernel.

Reply:
Do we need high precision timer? I think if `tick` is millsecond, use `now_coarse` is ok, or we need `std::time::Instant::now()`.

Reply:
Just a reminder. :wink: I recommend `Instant::now()` if it is not invoked very frequently.